validator/beacon/service.go

```diff
@@ -131,88 +184,58 @@ func (s *Service) fetchCrystallizedState(client pb.BeaconServiceClient) {
-			continue
+		if err := s.processCrystallizedState(crystallizedState); err != nil {
+			log.Error(err)
 		}
-		var crystallizedStateHash [32]byte
-		h := blake2b.Sum512(stateData)
-		copy(crystallizedStateHash[:], h[:32])
-
-		dynasty := crystallizedState.GetCurrentDynasty()
+	}
+}
 
-		for i, validator := range crystallizedState.GetValidators() {
-			if validator.StartDynasty <= dynasty && dynasty < validator.EndDynasty {
-				activeValidatorIndices = append(activeValidatorIndices, i)
-			}
-		}
-		isValidatorIndexSet := false
-		for _, val := range activeValidatorIndices {
-			// TODO: Check the public key instead of withdrawal address. This will use BLS.
-			if isZeroAddress(crystallizedState.Validators[val].WithdrawalAddress) {
-				s.validatorIndex = val
-				isValidatorIndexSet = true
-				break
-			}
-		}
+// processCrystallizedState uses a received crystallized state to determine
+// whether a validator is a proposer/attester and the validator's assigned slot.
+func (s *Service) processCrystallizedState(crystallizedState *pbp2p.CrystallizedState) error {
+	var activeValidatorIndices []int
+	dynasty := crystallizedState.GetCurrentDynasty()
 
-		// If validator was not found in the validator set was not set, keep listening for
-		// crystallized states.
-		if !isValidatorIndexSet {
-			log.Debug("Validator index not found in latest crystallized state's active validator list")
-			continue
+	for i, validator := range crystallizedState.GetValidators() {
+		if validator.StartDynasty <= dynasty && dynasty < validator.EndDynasty {
+			activeValidatorIndices = append(activeValidatorIndices, i)
 		}
+	}
+	isValidatorIndexSet := false
 
-		req := &pb.ShuffleRequest{
-			CrystallizedStateHash: crystallizedStateHash[:],
+	// We then iteratate over the activeValidatorIndices to determine what index
+	// this running validator client corresponds to.
+	for _, val := range activeValidatorIndices {
+		// TODO: Check the public key instead of withdrawal address. This will use BLS.
+		if isZeroAddress(crystallizedState.Validators[val].WithdrawalAddress) {
+			s.validatorIndex = val
+			isValidatorIndexSet = true
+			break
 		}
+	}
 
-		res, err := client.FetchShuffledValidatorIndices(s.ctx, req)
-		if err != nil {
-			log.Errorf("Could not fetch shuffled validator indices: %v", err)
-			continue
-		}
+	// If validator was not found in the validator set was not set, keep listening for
+	// crystallized states.
+	if !isValidatorIndexSet {
+		log.Debug("Validator index not found in latest crystallized state's active validator list")
+		return nil
+	}
 
-		shuffledIndices := res.GetShuffledValidatorIndices()
-		if uint64(s.validatorIndex) == shuffledIndices[len(shuffledIndices)-1] {
-			// The validator needs to propose the next block.
-			s.responsibility = "proposer"
-			log.Debug("Validator selected as proposer of the next slot")
-			continue
-		}
+	// TODO: Go through each of the indices for slots and determine which slot
```

Comment:
NOTE: I will implement this functionality leveraging the RPC methods @terenc3t defined in #526. The diff is already pretty big so I'd rather do it in a different PR.